File: app/modules/system_status/schemas.py
from __future__ import annotations

from datetime import datetime

from app.core.openai.status_state import OPENAI_STATUS_PAGE_URL
from app.modules.shared.schemas import DashboardModel


class SystemStatusComponentResponse(DashboardModel):
    name: str
    status: str


class SystemStatusIncidentResponse(DashboardModel):
    name: str
    impact: str
    status: str
    started_at: datetime | None = None
    shortlink: str | None = None
    affected_components: list[str] = []


class SystemStatusResponse(DashboardModel):
    # ``indicator`` is the Statuspage overall indicator (none/minor/major/critical)
    # or "unknown" when the proxy has not fetched a snapshot yet.
    indicator: str
    description: str
    components: list[SystemStatusComponentResponse]
    incidents: list[SystemStatusIncidentResponse]
    updated_at: datetime | None
    stale: bool
    status_page_url: str = OPENAI_STATUS_PAGE_URL

File: frontend/src/features/system-status/schemas.ts
import { z } from "zod";

export const SystemStatusComponentSchema = z.object({
  name: z.string(),
  status: z.string(),
});

export const SystemStatusIncidentSchema = z.object({
  name: z.string(),
  impact: z.string(),
  status: z.string(),
  startedAt: z.string().nullable().optional(),
  shortlink: z.string().nullable().optional(),
  affectedComponents: z.array(z.string()),
});

export const SystemStatusSchema = z.object({
  indicator: z.string(),
  description: z.string(),
  components: z.array(SystemStatusComponentSchema),
  incidents: z.array(SystemStatusIncidentSchema),
  updatedAt: z.string().nullable().optional(),
  stale: z.boolean(),
  statusPageUrl: z.string(),
});

export type SystemStatus = z.infer<typeof SystemStatusSchema>;
export type SystemStatusIncident = z.infer<typeof SystemStatusIncidentSchema>;

export const DEGRADED_INDICATORS = new Set(["minor", "major", "critical"]);
